### ISSUE TITLE
Notify changes on an item

### ISSUE THREAD

**Issue by folivares** [wasabeef/recyclerview-animators#35]
Hi, nice library! 
The animation on insert, delete and show item work very well but when I call "notifyItemChanged" or "notifyDatasetChanged" on the adapter, nothing happens.

**Comment by ahornerr** [comment]
:+1: I noticed this issue too. When I wrap my adapter in an `AlphaInAnimatorAdapter` and a `ScaleInAnimatorAdapter`, calling `notifyItemRemoved()` and `notifyDatasetChanged` on my original adapter causes some weirdness. Removing the adapter animation wrappers makes the issue disappear.

**Comment by wasabeef** [comment]
@ahornerr 

thanks. I will fix.

**Comment by ahornerr** [comment]
I believe this is a duplicate of #31

**Comment by wasabeef** [comment]
@ahornerr 

thanks

**Comment by ahmedlhanafy** [comment]
So is there any progress in this bug because it's stopping me from using the library

**Comment by wasabeef** [comment]
@folivares 
@ahornerr 
@ahmedlhanafy 

`notifyDataSetChanged` as a standard behavior dose not animation.
Please use the `notifyItemRemoved` and `notifyItemInserted`.

**Comment by roshu10** [comment]
@wasabeef hi, Thanks for the nice library. I tried to use "notifyItemRemoved" and "notifyItemInserted"
Both are not working

**Comment by craya1982** [comment]
It's hard to debug with out code. Are you expecting this behavior the very first time your recylcer view is drawn?  If so, it won't animate on first draw.  You need to delay adding items to your data set then notifying the adapter that you inserted item at x position.     Could you post some sample code?

**Comment by roshu10** [comment]
oncreateview(....) {
        View rootView = inflater.inflate(R.layout.fragment_work, container, false);

        // TodoDatabaseHandler is a SQLiteOpenHelper class connecting to SQLite
        dbHelper = new DatabaseHelper(getActivity());

        // Get access to the underlying writeable database
        SQLiteDatabase db = dbHelper.getWritableDatabase();

        swipeRefreshLayout = (SwipeRefreshLayout)  rootView.findViewById(R.id.swipe_refresh_layout);
        swipeRefreshLayout.setOnRefreshListener(this);
       
        Cursor todoCursor = db.rawQuery("SELECT  * FROM TABLE_WORK ORDER BY _ID DESC", null);

        mRecyclerView = (RecyclerView) rootView.findViewById(R.id.lvItems);
        mRecyclerView.setHasFixedSize(true);

        // use a linear layout manager
        mLayoutManager = new LinearLayoutManager(getActivity());
        mRecyclerView.setLayoutManager(mLayoutManager);

        // Setup cursor adapter using cursor from last step
        todoAdapter = new MyListCursorAdapter(getActivity(), todoCursor);


        alphaAdapter = new ScaleInAnimationAdapter(todoAdapter);
        alphaAdapter.setFirstOnly(false);
        alphaAdapter.setDuration(500);
        alphaAdapter.setInterpolator(new OvershootInterpolator());

        // Attach cursor adapter to the ListView
        mRecyclerView.setAdapter(alphaAdapter);

}

    @Override
    public void onRefresh() {
        
        // I made both data and wrapper adapter global
        todoAdapter.notifyDataSetChanged();
        alphaAdapter.notifyDataSetChanged();

}

**Comment by ahornerr** [comment]
@craya1982 @wasabeef The issue I'm having exists when using `notifyItemRemoved` and `notifyItemInserted`. When these methods are executed, the RecyclerView does nothing. Removing the wrapper adapters makes it work again. It seems that the AnimationAdapters are not passing these events through to the base adapter.

**Comment by roshu10** [comment]
Many guys mentioned the same in few issues, Is that resolved ?

On Sun, Sep 20, 2015 at 10:44 PM, Andy Horner <notifications@github.com>
wrote:

> @craya1982 <https://github.com/craya1982> @wasabeef
> <https://github.com/wasabeef> The issue I'm having exists when using
> notifyItemRemoved and notifyItemInserted. When these methods are
> executed, the RecyclerView does nothing. Removing the wrapper adapters
> makes it work again. It seems that the AnimationAdapters are not passing
> these events through to the base adapter.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/wasabeef/recyclerview-animators/issues/35#issuecomment-141811683>
> .
>



-- 
Regards,
Roshan


*The only way to do great work is to love what you do. If you haven’t found
it yet, keep looking. Don’t settle. As with all matters of the heart,
you’ll know when you find it.” - Steve Jobs (1955 - 2011*

**Comment by anidotnet** [comment]
I am also facing the similar issue as @ahornerr. When I am not using AnimatorAdapter RecyclerView updates itself with new data, but when I wrap the cursor adapter with AnimatorAdapter, RecyclerView does not update itself. Please fix this bug.

**Comment by anidotnet** [comment]
Guys, I got it working somehow. Here is the code.

```java
AlphaInAnimationAdapter alphaAdapter = new AlphaInAnimationAdapter(mAdapter);
final ScaleInAnimationAdapter scaleAdapter = new ScaleInAnimationAdapter(alphaAdapter);
scaleAdapter.setFirstOnly(false);
scaleAdapter.setInterpolator(new OvershootInterpolator());
mRecyclerView.setItemAnimator(new FlipInRightYAnimator());
mRecyclerView.getItemAnimator().setAddDuration(300);

mRecyclerView.setAdapter(scaleAdapter);
```
And from background thread, I am calling

```java
parentActivity.getLoaderManager().restartLoader(0, null, MyFragment.this);
// size is the number of elements you inserted.
scaleAdapter.notifyItemRangeInserted(0, size);  
```
RecyclerView now loads the data automatically.

**Comment by wasabeef** [comment]
Please use the following  
`notifyItemChanged(int)`
`notifyItemInserted(int)`
`notifyItemRemoved(int)`
`notifyItemRangeChanged(int, int)`
`notifyItemRangeInserted(int, int)`
`notifyItemRangeRemoved(int, int)`

If you want your animations to work, do not rely on calling `notifyDataSetChanged()`; 
as it is the RecyclerView's default behavior, animations are not triggered to start inside this method.

**Comment by ahornerr** [comment]
@wasabeef 

I noticed the issue happening with `notifyItemRemoved(int)`

**Comment by eosrei** [comment]
FWIW Animations work in the stock RecyclerView with `notifyDataSetChanged()` if `HasStableIds` is *true* and `getItemId()` is used.

> Just set the following in activity,

> `recyclerView.setHasFixedSize(true);`

> And in adapter, write
> ```
> setHasStableIds(true); //in constructor 
> @Override
>     public long getItemId(int position) {
>         return cameraImageArrayList.get(position).hashCode(); //Any unique id
>     }
> ```
src: http://stackoverflow.com/a/31697304/1044864

**Comment by Piasy** [comment]
@wasabeef @ahornerr  
I think the root cause is that [this method is overrided](https://github.com/wasabeef/recyclerview-animators/blob/9bd8ea106b3bc4d7b4b6736c88a2efdafaae638c/animators%2Fsrc%2Fmain%2Fjava%2Fjp%2Fwasabeef%2Frecyclerview%2Fanimators%2FBaseItemAnimator.java#L328), if we call `notifyItemChanged(int)`, `animateChange(ViewHolder oldHolder, ViewHolder newHolder, ItemHolderInfo preLayoutInfo, ItemHolderInfo postLayoutInfo)` will be called, and this method should call `dispatchAnimationFinished(ViewHolder)` finally, which can let RecyclerView knows that the animation is ended. 

But in this wrong overriding, it just returns `false`, and called neither `super.animateChange(ViewHolder oldHolder, ViewHolder newHolder, ItemHolderInfo preLayoutInfo, ItemHolderInfo postLayoutInfo)`, nor `dispatchAnimationFinished(ViewHolder)`, which is definitely wrong.

Because it doesn't call super, [this overriding](https://github.com/wasabeef/recyclerview-animators/blob/9bd8ea106b3bc4d7b4b6736c88a2efdafaae638c/animators%2Fsrc%2Fmain%2Fjava%2Fjp%2Fwasabeef%2Frecyclerview%2Fanimators%2FBaseItemAnimator.java#L334) will never be called, because it is only called by `SimpleItemAnimator`.

So this bug will cause two problems:
1. change animation will never be played correctly;
2. once an item view triggered change animation, other animations on this item will be affected too.

Simply not overriding this method will fix this bug, I have submitted a pr #79 to fix it, hope @wasabeef process it soon. :)

**Comment by wasabeef** [comment]
@Piasy @ahornerr 

Released `2.2.2`

**Comment by Piasy** [comment]
Great!
